apps/aesophia/src/aeso_builtins.erl

```diff
@@ -0,0 +1,481 @@
+
+builtin_deps1({map_lookup_default, Type}) -> [{map_lookup, Type}];
+builtin_deps1({map_get, Type})            -> [{map_lookup, Type}];
+builtin_deps1(map_member)                 -> [{map_lookup, word}];
+builtin_deps1({map_upd, Type})            -> [{map_get, Type}, map_put];
+builtin_deps1({map_upd_default, Type})    -> [{map_lookup_default, Type}, map_put];
+builtin_deps1(map_from_list)              -> [map_put];
+builtin_deps1(str_equal)                  -> [str_equal_p];
+builtin_deps1(string_concat)              -> [string_concat_inner1, string_concat_inner2];
+builtin_deps1(int_to_str)                 -> [int_to_str_, int_digits];
+builtin_deps1(addr_to_str)                -> [base58_int, string_concat];
+builtin_deps1(base58_int)                 -> [base58_int_encode, base58_int_pad, string_reverse, string_concat];
+builtin_deps1(base58_int_encode)          -> [base58_int_encode_, base58_tab];
+builtin_deps1(string_reverse)             -> [string_reverse_];
+builtin_deps1(_)                          -> [].
+
+dep_closure(Deps) ->
+    case lists:umerge(lists:map(fun builtin_deps/1, Deps)) of
+        []    -> Deps;
+        Deps1 -> lists:umerge(Deps, dep_closure(Deps1))
+    end.
+
+%% Helper functions/macros
+v(X) when is_atom(X) -> v(atom_to_list(X));
+v(X) when is_list(X) -> #var_ref{name = X}.
+
+option_none()  -> {tuple, [{integer, 0}]}.
+option_some(X) -> {tuple, [{integer, 1}, X]}.
+
+-define(call(Fun, Args), #funcall{ function = #var_ref{ name = {builtin, Fun} }, args = Args }).
+-define(I(X), {integer, X}).
+-define(V(X), v(X)).
+-define(A(Op), aeb_opcodes:mnemonic(Op)).
+-define(LET(Var, Expr, Body), {switch, Expr, [{v(Var), Body}]}).
+-define(DEREF(Var, Ptr, Body), {switch, v(Ptr), [{{tuple, [v(Var)]}, Body}]}).
+-define(NXT(Ptr), op('+', Ptr, 32)).
+-define(NEG(A), op('/', A, {unop, '-', {integer, 1}})).
+-define(BYTE(Ix, Word), op('byte', Ix, Word)).
+
+-define(EQ(A, B), op('==', A, B)).
+-define(LT(A, B), op('<', A, B)).
+-define(GT(A, B), op('>', A, B)).
+-define(ADD(A, B), op('+', A, B)).
+-define(SUB(A, B), op('-', A, B)).
+-define(MUL(A, B), op('*', A, B)).
+-define(DIV(A, B), op('div', A, B)).
+-define(MOD(A, B), op('mod', A, B)).
+-define(EXP(A, B), op('^', A, B)).
+-define(AND(A, B), op('&&', A, B)).
+
+-define(BSL(X, B), ?MUL(X, ?EXP(2, ?MUL(B, 8)))).
+-define(BSR(X, B), ?DIV(X, ?EXP(2, ?MUL(B, 8)))).
+
+op(Op, A, B) -> {binop, Op, operand(A), operand(B)}.
+
+operand(A) when is_atom(A) -> v(A);
+operand(I) when is_integer(I) -> {integer, I};
+operand(T) -> T.
+
+str_to_icode(String) when is_list(String) ->
+    str_to_icode(list_to_binary(String));
+str_to_icode(BinStr) ->
+    Cpts = [size(BinStr) | aeso_data:binary_to_words(BinStr)],
+    #tuple{ cpts = [ #integer{value = X} || X <- Cpts ] }.
+
+%% Event primitive (dependent on Event type)
+%%
+%% We need to switch on the event and prepare the correct #event for icode_to_asm
+%% NOTE: we assume all errors are already checked!
```

Comment:
Are they though? Somewhere we should check that indexed arguments are words and that the payload is a string. This is a nice place to do it since we can translate Sophia types to VM types here.

Reply:
Does Robert agree this is a nice place? 😉 

Any pointers on how to translate Sophia types to VM types would be helpful?! How to report an error from here?